oscore-cf/src/main/java/org/eclipse/californium/oscore/ResponseDecryptor.java

```diff
@@ -0,0 +1,79 @@
+/*******************************************************************************
+ * Copyright (c) 2018 RISE SICS and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Joakim Brorsson
+ *    Ludwig Seitz (RISE SICS)
+ *    Tobias Andersson (RISE SICS)
+ *    
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import java.util.Arrays;
+import java.util.logging.Logger;
+
+import org.eclipse.californium.core.coap.CoAP;
+import org.eclipse.californium.core.coap.Response;
+import org.eclipse.californium.core.coap.Token;
+
+public class ResponseDecryptor extends Decryptor {
+
+	/**
+	 * The logger
+	 */
+	private static final Logger LOGGER = Logger.getLogger(ResponseDecryptor.class.getName());
+
+	/**
+	 * Decrypt the response.
+	 * 
+	 * @param response the response
+	 * @param db the database of OSCore contexts
+	 * 
+	 * @return the decrypted response
+	 * 
+	 * @throws OSException when decryption fails
+	 * 
+	 */
+	public static Response decrypt(Response response) throws OSException {
+
+		discardEOptions(response);
+
+		OSCoreCtxDB db = HashMapCtxDB.getInstance();
+
+		byte[] protectedData = collectData(response);
+		Encrypt0Message enc = null;
+		Token token = response.getToken();
+		OSCoreCtx ctx = null;
+
+		if (token != null) {
+			ctx = db.getContextByToken(token);
+			if (ctx == null) {
+				LOGGER.severe(OSExceptionHandler.TOKEN_INVALID);
+				throw new OSException(OSExceptionHandler.TOKEN_INVALID);
+			}
+			enc = decompression(protectedData, response);
+		} else {
+			LOGGER.severe(OSExceptionHandler.TOKEN_NULL);
+			throw new OSException(OSExceptionHandler.TOKEN_NULL);
+		}
+
+		byte[] content = decryptAndDecode(enc, response, ctx, db.getSeqByToken(token));
```

Comment:
with PR #660 it should look similar to:

```
byte[] plaintext = decryptAndDecode(enc, response, ctx, db.getSeqByToken(token));
DatagramReader reader = new DatagramReader(new ByteArrayInputStream(plaintext));
response = OptionJuggle.setRealCodeResponse(response, CoAP.ResponseCode.valueOf(reader.read(CoAP.MessageFormat.CODE_BITS)));
DataParser.parseOptionsAndPayload(reader, response);

```

The old implementation seems to ignore E options for responses. 
This new old implementation will not mix together the E and U options into the original options. 
So I guess, there is still something missing.

Reply:
Yes, I think you are correct! The ResponseDecryptor is missing setting the E options in my code for some reason. The code doesn't even read the E options from the content array. How can this code have worked? It is a huge oversight from my side.